File: bin/go/witch.go
package main

import (
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strings"

	"github.com/fatih/color"
)

const doClientMode = "WITCH_CLIENT_MODE"

func main() {
	if len(os.Args) > 2 && os.Args[1] == doClientMode {
		clientMode(os.Args[2], os.Args[3:])
	} else {
		entrMode()
	}
}

func entrMode() {
	if len(os.Args) < 3 {
		fatal("usage: witch <glob> <command> [args...]")
	}
	self := os.Args[0]
	glob := os.Args[1]
	command := os.Args[2:]
	globs, err := filepath.Glob(glob)
	check(err)
	cmd := exec.Command("entr", append([]string{"-c", self, doClientMode}, command...)...)
	cmd.Stdin = strings.NewReader(strings.Join(globs, "\n"))
	cmd.Stdout = os.Stdout
	cmd.Stderr = os.Stderr
	checkExit(cmd.Run())
}

func clientMode(command string, args []string) {
	cmd := exec.Command(command, args...)
	cmd.Stdin = os.Stdin
	cmd.Stdout = os.Stdout
	cmd.Stderr = os.Stderr
	checkExit(cmd.Run())
	color.Green("ok")
}

func checkExit(err error) {
	if eErr, ok := err.(*exec.ExitError); ok && eErr.ExitCode() != 0 {
		os.Exit(eErr.ExitCode())
	}
	check(err)
}

func check(err error) {
	if err != nil {
		fatal(err)
	}
}

func fatal(v interface{}) {
	fmt.Fprintln(os.Stderr, v)
	os.Exit(1)
}

File: scripts/witch.py
# {"deps": ["entr"], "requirements": ["termcolor"]} #nix
from subprocess import run
import glob
import sys

from termcolor import cprint

do_client_mode = 'WITCH_CLIENT_MODE'
argv = sys.argv

def main():
    try:
        if len(argv) > 2 and argv[1] == do_client_mode:
            client_mode(argv[2:])
        else:
            entr_mode()
    except KeyboardInterrupt:
        pass

def entr_mode():
    if len(argv) < 3:
        print('usage: witch <glob> <command> [args...]', file=sys.stderr)
        sys.exit(1)
    self, path_glob, command = argv[0], argv[1], argv[2:]
    entr = run(
        ['entr', '-c', sys.executable, self, do_client_mode] + command,
        input='\n'.join(glob.glob(path_glob, recursive=True)).encode(),
    )
    if entr.returncode != 0:
        sys.exit(entr.returncode)

def client_mode(cmd):
    code = run(cmd).returncode
    if code == 0:
        cprint('ok', 'green')
    else:
        cprint('failed', 'red')
        sys.exit(code)

if __name__ == '__main__':
    main()
